### PR TITLE
Add module reader to enable module dependent variables

### PR DESCRIPTION
**Pull request by Nepomuceno**
This Pr it is part of a larger PR it order to add variable auto completion for modules in the terraform language server. https://github.com/hashicorp/terraform-ls/issues/93
a complementary implementation in the language server side can be found at 
https://github.com/hashicorp/terraform-ls/pull/551

### CODE REVIEW

**Review by radeksimko — COMMENTED** [review]
Except for my in-line comments (most of which are minor) this looks good!

Can you also add a new test to `schema/schema_merge_test.go` which exercises the new codepath? i.e. has a simple mock implementation of the `ModuleReader` interface (can be literally hard-coded return values) and checks that the `module` block has the expected schema?

Thanks!

File: schema/schema_merge.go [hunk 1]
```diff
@@ -9,18 +9,26 @@ import (
...
 	"github.com/hashicorp/terraform-schema/internal/schema/backends"
 	"github.com/hashicorp/terraform-schema/module"
+	"github.com/zclconf/go-cty/cty"
 )
 
 type SchemaMerger struct {
 	coreSchema       *schema.BodySchema
 	schemaReader     SchemaReader
 	terraformVersion *version.Version
+	moduleReader     *ModuleReader
```
[the remark this answers is not in the document]
Comment (Nepomuceno) [reply schema/schema_merge.go:19]: The problem with this is how can I know that I have a module reader added I wanted to test if it exists before actually calling if with an interface I cannot check for nil. If you know a way for me to check if it has been declared before trying to access without using a pointer that would be great. :) Will see if I can find how to do it online.

File: schema/schema_merge.go [hunk 2]
```diff
@@ -194,6 +210,34 @@ func (m *SchemaMerger) SchemaForModule(meta *module.Meta) (*schema.BodySchema, e
 		}
 		mergedSchema.Blocks["variable"].DependentBody = variableDependentBody(meta.Variables)
 	}
+	if m.moduleReader != nil {
+		reader := *m.moduleReader
```
[the remark this answers is not in the document]
Comment (Nepomuceno) [reply schema/schema_merge.go:214]: That is the line that made me use an interface I just want to execute this code if the module reader had been set.

File: schema/schema_merge.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply schema/schema_merge.go:19]: You can compare the interface to `nil` AFAIK - that's what we do with `SchemaReader` already. I admit I don't know all the "why"s though (i.e. what exactly happens behind the scenes in the compiler and why it works the way it works).

[the remark this answers is not in the document]
Comment (Nepomuceno) [reply schema/schema_merge.go:19]: Apparently you can just compare an interface to nil ... Now I am curious to understand how this works...

**Review by radeksimko — APPROVED** [review]
LGTM, I think this is ready for merging, but before doing so I'd like to also see https://github.com/hashicorp/terraform-ls/pull/551 modified to leverage this PR - so we're sure that everything fits well together.

**Comment by Nepomuceno** [comment]
> LGTM, I think this is ready for merging, but before doing so I'd like to also see [hashicorp/terraform-ls#551](https://github.com/hashicorp/terraform-ls/pull/551) modified to leverage this PR - so we're sure that everything fits well together.

Working on that now just wanted to make sure this interface would not change.

**Pull request merged**